lib/web/app/transport.go

```diff
@@ -0,0 +1,215 @@
+	cookies := r.Cookies()
+	r.Header.Del("Cookie")
+	for _, cookie := range cookies {
+		if cookie.Name == cookieName {
+			continue
+		}
+		r.AddCookie(cookie)
+	}
+
+	return nil
+}
+
+// dialFunc returns a function that can Dial and connect to the application
+// service over the reverse tunnel subsystem.
+func dialFunc(c *transportConfig) func(ctx context.Context, network string, addr string) (net.Conn, error) {
+	return func(ctx context.Context, network string, addr string) (net.Conn, error) {
+		clusterClient, err := c.proxyClient.GetSite(c.identity.RouteToApp.ClusterName)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		conn, err := clusterClient.Dial(reversetunnel.DialParams{
+			// The "From" and "To" addresses are not actually used for tunnel dialing,
+			// they're filled out with "dummy values" that make logs easier to read and
+			// debug within the reverse tunnel subsystem.
+			From:     &utils.NetAddr{AddrNetwork: "tcp", Addr: "@web-proxy"},
+			To:       &utils.NetAddr{AddrNetwork: "tcp", Addr: fmt.Sprintf("@app-%v", c.app.Name)},
+			ServerID: fmt.Sprintf("%v.%v", c.server.GetName(), c.identity.RouteToApp.ClusterName),
+			ConnType: services.AppTunnel,
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return conn, nil
+	}
+}
+
+// configureTLS creates and configures a *tls.Config will be used for
+// mutual authentication.
+func configureTLS(c *transportConfig) (*tls.Config, error) {
+	tlsConfig := utils.TLSConfig(c.cipherSuites)
+
+	// Configure the pool of certificates that will be used to verify the
+	// identity of the server. This allows the client to verify the identity of
+	// the server it is connecting to.
+	ca, err := c.accessPoint.GetCertAuthority(services.CertAuthID{
+		Type:       services.HostCA,
+		DomainName: c.identity.RouteToApp.ClusterName,
+	}, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	certPool, err := services.CertPool(ca)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tlsConfig.RootCAs = certPool
+
+	// Configure the identity that will be used to connect to the server. This
+	// allows the server to verify the identity of the caller.
+	certificate, err := tls.X509KeyPair(c.ws.GetTLSCert(), c.ws.GetPriv())
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse certificate or key")
+	}
+	tlsConfig.Certificates = []tls.Certificate{certificate}
+
+	// Make sure the server this client connects to presents a certificate for
+	// hostUUID.clusterName.
+	tlsConfig.ServerName = fmt.Sprintf("%v.%v", c.server.GetName(), c.identity.RouteToApp.ClusterName)
+
+	// This is a hack used elsewhere within Teleport to ensure that a client
+	// always sends it's clients certificate when establishing a connection. It's
+	// unclear if it's still needed in recent versions of Go.
+	cert := tlsConfig.Certificates[0]
+	tlsConfig.Certificates = nil
+	tlsConfig.GetClientCertificate = func(_ *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		return &cert, nil
+	}
```

Comment:
@fspmarshall found this: https://github.com/golang/go/issues/23924 
Looks like the problem is only when server sends algo constraints in certificate request, and the client cert doesn't match those. For example, server says "I'll only accept ECDSA client certs" and client has an RSA cert.

IIUC, since we control both ends of the connection, this workaround can be removed.